### DISCUSSION TITLE
Integration of the Microsoft WebView2 control.

### DISCUSSION THREAD

**Discussion by laultman** [StereoKit/StereoKit#1057] [Q&A]
@maluoi I've been working on this for a couple days. First question, do you know if anyone has it working in StereoKit? Observation, the documentation from Microsoft leaves a lot to be desired - that's really nothing new.
I've successfully gotten it to initialize in a standard console app that does not have a winform by creating a window handle. WebView2 requires a HWND that is normally the UI thread. So far, I can't coax SK into giving me the UI thread handle required. The WebView2 control is closely aligned with Microsoft's WinUI-2 and the coming WinUI-3. These do nothing for me in SK. Microsoft recommends using the composition mode for use cases such as SK. I'm sure that will be a challenge but so far I still I'm stuck at the initialization.
By the way I tried using the SK.ExecuteOnMain() but that has not worked. ANY suggestions would be appreciated!

**Comment by maluoi** [comment]
Yeah, the UWP window does run on a separate thread from StereoKit, so SK's thread is not what you'd need for that! In native code, I use `CoreApplication::MainView().CoreWindow().Dispatcher().RunAsync()` to execute code on the UWP UI thread, I expect there's a C# equivalent of that you could use to run your code there?

**Comment by laultman** [comment]
@maluoi I decided to start by injecting the WebView2 (WV2) control directly into my project leaving everything SK as it currently is. Could not get it to initialize. Turns out that the necessary libraries were not being installed, just a loader. Something to do with the way WV2 uses a runtime loader which UWP seems to barf over. I tried using the native stuff from the Unity implementation to try and initialize using native WV2 calls. I could get the calls to work but it always failed with a runtime UI threading error.

Next I added the nuget packages 2 required (WV2 and the Microsoft.Windows.ImplementationLibrary) to the StereoKitC_UWP. This compiled fine and the necessary libs were in the output but the issue with runtime UI threading still prevented me from initializing the control.

Now after a coupled days of work in the project XML of both the StereoKitC and StereoKitC_UWP I get good builds of the SK with all the WV2 now residing directly in the StereoKitC. All the projects build and run with the inclusion of the WV2. It is not yet being called, it's just there. Next step is initialization which is to try and initialize immediately after you initialize the SK UI.

Any suggestions here would be helpful. Attached are my modified StereoKitC and StereoKitC_UWP project files. Note, I am developing on an ARM64 equipped machine, so I haven't tested on x64. I added to both C++ projects nuget packages:  WebView2 and Microsoft.Windows.ImplementationLibrary. Note: Be careful to add the same versions to both projects!
Finally, I updated the nuget package for Microsoft.Windows.CppWinRT.2.0.240405.15. I had to copy the version number into the existing reference line and delete the one inserted by the package manager because you are using a non-local path to the dll files. This is the modified line:
`  <Import Project="$(ProjectDir)..\..\packages\Microsoft.Windows.CppWinRT.2.0.240405.15\build\native\Microsoft.Windows.CppWinRT.props" Condition="Exists('$(ProjectDir)..\..\packages\Microsoft.Windows.CppWinRT.2.0.240405.15\build\native\Microsoft.Windows.CppWinRT.props')" />
`

[StereoKitC.zip](https://github.com/user-attachments/files/16402765/StereoKitC.zip)

**Comment by laultman** [comment]
@maluoi Beginning the WebView2 control initialization in the SK C++ project. I am using the Microsoft WebView2 control sample C++ with the composition host. After reading lots of MS documentation :) - it seems that for projects **NOT** using the Windows frameworks UI (WinUI, WPF, UWP/XAML) you must use a native composition host. I'm sure this going to be difficult since I know almost nothing about swapchain stuff.

As a beginning though, I have borrowed a couple files and a couple .h from the project sample and modified them a bit to get a good build. I added a new folder to the StereoKitC project, WebView2 and linked it into the project as you have your other folders. In it I put the .h and .cpp files. At present the projects build and the test project runs as expected. I still have not done the initialization, just added some code for a start.

Attached is my WebView2 folder
[webview2.zip](https://github.com/user-attachments/files/16402850/webview2.zip)
2 folder and files - does nothing except exist in the project.

**Comment by laultman** [comment]
@maluoi I could not get the StereoKitCTest to run in debug mode on ARM64 without a lot of work on the project and build cmake so I switched to an X64 machine to do the WebView2 C++ implementation. FYI StereoKitCTest runs fine on X64 in debug mode.

**Comment by laultman** [comment]
@maluoi I've finally gotten WebView2 control to work in a C++ project with OpenXR. It took a lot of work to figure out all the configurations int VS C++ and then resolve all the missing dependencies. Next step is Windows Events and swapchain. It loads to HoloLens 2 and runs until it hits one of the many procedures that I am not yet handling.
